### DISCUSSION TITLE
求教 当组件属性中存在array或者object类型的属性时，会导致TreeNodeWidget组件渲染两次或者三次的问题

### DISCUSSION THREAD

**Discussion by superchenwb** [alibaba/designable#36] [Q&A] [answered]
当组件属性中存在array或者object类型的属性时，拖拽组件到面板的时候，会导致TreeNodeWidget组件渲染两次，如果没有object或array类型的属性时，只会渲染一次

添加一个array属性
![image](https://user-images.githubusercontent.com/17635660/128618743-5be31de1-7a19-4e4f-8b7e-1572f863e837.png)
设置一个默认值
![image](https://user-images.githubusercontent.com/17635660/128618791-46daf99e-6aa9-4ca7-9dc8-fe648692585f.png)
进行打印
![image](https://user-images.githubusercontent.com/17635660/128618755-4fc569a3-d161-4d67-a155-36389ea42001.png)
添加默认值渲染两次
![image](https://user-images.githubusercontent.com/17635660/128618768-c01e87ac-80a6-4c48-9840-0886c58f9c38.png)
不添加默认值，会渲染三次
![image](https://user-images.githubusercontent.com/17635660/128618842-6f65f0b3-6676-4d72-b49f-9feed4845133.png)

**Comment by janryWang** [comment]
排除法定位问题吧，把问题范围缩到最小才方便确定问题原因

**Reply by superchenwb** [reply]
![image](https://user-images.githubusercontent.com/17635660/128622044-b252206d-a3a4-44e2-94dd-875677b46d05.png)
将第35行去掉，就不会重复渲染了。
我猜测应该是formily表单在渲染完成之后，对传入的值进行observable处理，导致node的props属性的对象或数组引用发生了变化。
于是发生了重复渲染，毕竟如果是基础数据类型，不会出现这个问题

我注意到项目中对于style对象的处理，是使用style.width之类的属性作为key，单独拆出来的，或许当初就是为了规避这个问题
![image](https://user-images.githubusercontent.com/17635660/128622165-9c8b779f-1543-4098-b92b-6c1446416052.png)
于是我修改了一下style属性，修改为嵌套对象
![image](https://user-images.githubusercontent.com/17635660/128622286-288d2e55-c2fc-4c1d-8fce-a8b873a90e43.png)
结果发现，死循环了
![image](https://user-images.githubusercontent.com/17635660/128622281-4bc76a7c-2774-4c60-9245-ce3f9eb85e0d.png)

内存占用达到了700多MB
![image](https://user-images.githubusercontent.com/17635660/128623090-abd3eec0-a936-46b2-8e3c-ffda43061127.png)

**Comment by janryWang** [comment]
看了一下，渲染2次或3次是合理的，因为对于数组或者对象，formily是会做默认值修正的，这里面是会存在一次数据操作，所以如果你没有设置默认值，那肯定会渲染两次，然后如果你设置了默认值，那就会渲染1次，但是问题来了，为啥最低是渲染两次。。。

这是因为SettingsForm内部依赖了schema变化，恰好Field的propsSchema是函数，内部依赖了节点componentName，所以拖进去的时候，schema会发生变化，所以会触发SettingsForm渲染两次，也就是会做两次数据修正

**Reply by superchenwb** [reply]
我不太理解拖进去的时候，为啥propsSchema会发生变化。

我又测试了一下，总结的规律如下：

1. 属性是数组，例如如上图所示添加titles数组属性
- 默认值为undefined，渲染3次
- 默认值为空数组[], 渲染2次
- 默认值不为空数组, 渲染1次

2. 属性是数组，例如如上图所示添加style对象属性
- 默认值为undefined，发生死循环
- 默认值为空对象{}, 发生死循环
- 默认值不为空对象style: {width: '200px'}, 渲染1次

发生死循环的原因我觉得不是schema发生了变化，而是因为SettingsForm组件使用了observer包裹，当存在对象属性时，formily对默认值进行了修正，导致node?.props发生率变化，进而SettingsForm组件重新渲染了，于是formily再次对默认值进行了修正，再次导致SettingsForm组件发生渲染，于是产生了死循环
（在SettingsForm打印调试语句，确实是会打印多次的）

以上的猜测能解释死循环的问题，确无法解释为什么是数组属性且不存在默认值时，确不会死循环，而是只渲染3次

**Comment by janryWang** [comment, marked as answer]
死循环的问题，我在最新版里已经修复了，渲染3次的问题是因为formily没对空对象或者空数组做取等判断，导致重复赋值，然后你上一个内存泄漏的问题，请录个复现视频出来，我这边再详细看看

**Reply by superchenwb** [reply]
谢谢啦，内存泄漏指的是我另外一个问题里提到的那个吗，那我待会发到那个帖子上去
